Implementation of the first performance test scenario

Dannyps opened the pull request:
# Readiness checklist

-   [ ] I added/updated unit tests.
-   [ ] I added/updated integration tests.
-   [x] I ensured that the PR title is good enough for the changelog.
-   [x] I labeled the PR.


Please refer to the added readme files:

https://github.com/nmshd/backbone/blob/NMSHDB-65-implementation-of-the-first-performance-test-scenario/Applications/ConsumerApi/test/ConsumerApi.Tests.Performance/src/scenarios/README.md

https://github.com/nmshd/backbone/blob/NMSHDB-65-implementation-of-the-first-performance-test-scenario/Applications/ConsumerApi/test/ConsumerApi.Tests.Performance/README.md

Example output of scenario **1** (verbose):
```
$ npx webpack && k6.exe run .\dist\s01.test.js -v
asset s01.test.js 70.7 KiB [emitted] (name: s01.test)
orphan modules 14.1 KiB [orphan] 11 modules
runtime modules 937 bytes 4 modules
built modules 66.5 KiB [built]
  ./src/scenarios/s01.test.ts + 11 modules 16.2 KiB [not cacheable] [built] [code generated]
  ./node_modules/papaparse/papaparse.js 50.3 KiB [built] [code generated]
  external "stream" 42 bytes [built] [code generated]
webpack 5.93.0 compiled successfully in 1755 ms
DEBU[0000] Logger format: TEXT
DEBU[0000] k6 version: v0.52.0 (commit/20f8febb5b, go1.22.4, windows/amd64)

          /\      |‾‾| /‾‾/   /‾‾/
     /\  /  \     |  |/  /   /  /
    /  \/    \    |     (   /   ‾‾\
   /          \   |  |\  \ |  (‾)  |
  / __________ \  |__| \__\ \_____/ .io

DEBU[0000] Resolving and reading test '.\dist\s01.test.js'...
DEBU[0000] Loading...                                    moduleSpecifier="file:///D:/backbone/Applications/ConsumerApi/test/PerformanceTests/dist/s01.test.js" originalModuleSpecifier=".\\dist\\s01.test.js"
DEBU[0000] '.\dist\s01.test.js' resolved to 'file:///D:/backbone/Applications/ConsumerApi/test/PerformanceTests/dist/s01.test.js' and successfully loaded 72349 bytes!
DEBU[0000] Gathering k6 runtime options...
DEBU[0000] Initializing k6 runner for '.\dist\s01.test.js' (file:///D:/backbone/Applications/ConsumerApi/test/PerformanceTests/dist/s01.test.js)...
DEBU[0000] Detecting test type for...                    test_path="file:///D:/backbone/Applications/ConsumerApi/test/PerformanceTests/dist/s01.test.js"
DEBU[0000] Trying to load as a JS test...                test_path="file:///D:/backbone/Applications/ConsumerApi/test/PerformanceTests/dist/s01.test.js"
DEBU[0000] Loading...                                    moduleSpecifier="https://jslib.k6.io/httpx/0.1.0/index.js" originalModuleSpecifier="https://jslib.k6.io/httpx/0.1.0/index.js"
DEBU[0000] Fetching source...                            url="https://jslib.k6.io/httpx/0.1.0/index.js?_k6=1"
DEBU[0000] Fetched!                                      len=8730 t=272.2943ms url="https://jslib.k6.io/httpx/0.1.0/index.js?_k6=1"
DEBU[0000] Babel: Transformed                            t=289.3635ms
INFO[0000] Loading identities                            source=console
INFO[0000] Finished Loading                              source=console
DEBU[0000] Runner successfully initialized!
DEBU[0000] Parsing CLI flags...
DEBU[0000] Consolidating config layers...
DEBU[0000] Parsing thresholds and validating config...
DEBU[0000] Initializing the execution scheduler...
DEBU[0000] Starting 2 outputs...                         component=output-manager
DEBU[0000] Starting...                                   component=metrics-engine-ingester
DEBU[0000] Started!                                      component=metrics-engine-ingester
     execution: local
        script: .\dist\s01.test.js
        output: -

     scenarios: (100.00%) 1 scenario, 1 max VUs, 1h0m30s max duration (incl. graceful stop):
              * constantRequestRate: 0.00 iterations/s for 1h0m0s (maxVUs: 1, gracefulStop: 30s)

DEBU[0000] Starting the REST API server on localhost:6565
DEBU[0000] Trapping interrupt signals so k6 can handle them gracefully...
DEBU[0000] Starting emission of VUs and VUsMax metrics...
DEBU[0000] Start of initialization                       executorsCount=1 neededVUs=1 phase=execution-scheduler-init
DEBU[0000] Initialized VU #1                             phase=execution-scheduler-init
DEBU[0000] Finished initializing needed VUs, start initializing executors...  phase=execution-scheduler-init
DEBU[0000] Initialized executor constantRequestRate      phase=execution-scheduler-init
DEBU[0000] Initialization completed                      phase=execution-scheduler-init
DEBU[0000] Start of test run                             executorsCount=1 phase=execution-scheduler-run
DEBU[0000] setup() is not defined or not exported, skipping!
DEBU[0000] Start all executors...                        phase=execution-scheduler-run
DEBU[0000] Starting executor                             executor=constantRequestRate startTime=0s type=constant-arrival-rate
DEBU[0000] Starting executor run...                      duration=1h0m0s executor=constant-arrival-rate maxVUs=1 preAllocatedVUs=1 scenario=constantRequestRate tickerPeriod=5m0s type=constant-arrival-rate
DEBU[3600] Regular duration is done, waiting for iterations to gracefully finish  executor=constant-arrival-rate gracefulStop=30s scenario=constantRequestRate
DEBU[3601] Executor finished successfully                executor=constantRequestRate startTime=0s type=constant-arrival-rate
DEBU[3601] teardown() is not defined or not exported, skipping!                                                                                                                                                                                 
DEBU[3601] Test finished cleanly                                                                                                                                                                                                                
DEBU[3601] Stopping vus and vux_max metrics emission...  phase=execution-scheduler-init                                                                                                                                                         
DEBU[3601] Sending usage report...                                                                                                                                                                                                              
DEBU[3601] Metrics emission of VUs and VUsMax metrics stopped                                                                                                                                                                                   
DEBU[3601] Releasing signal trap...                                                                                                                                                                                                             
DEBU[3601] Waiting for metrics and traces processing to finish...                                                                                                                                                                               
DEBU[3601] Metrics and traces processing finished!                                                                                                                                                                                              
DEBU[3601] Stopping outputs...                                                                                                                                                                                                                  
DEBU[3601] Stopping 2 outputs...                         component=output-manager                                                                                                                                                               
DEBU[3601] Stopping...                                   component=metrics-engine-ingester                                                                                                                                                      
DEBU[3601] Stopped!                                      component=metrics-engine-ingester                                                                                                                                                      
DEBU[3601] Generating the end-of-test summary...                                                                                                                                                                                                

     ✓ response code was 201
     ✓ challenge contains correct device
     ✓ challenge contains correct address
     ✓ challenge id is not empty

     checks.........................: 100.00% ✓ 52                  ✗ 0
     data_received..................: 28 kB   7.697837102571258 B/s
     data_sent......................: 18 kB   4.937836839599921 B/s
     http_req_blocked...............: avg=62.3µs   min=0s       med=0s       max=1.11ms   p(90)=0s       p(95)=381.75µs
     http_req_connecting............: avg=22.58µs  min=0s       med=0s       max=587.09µs p(90)=0s       p(95)=0s
     http_req_duration..............: avg=268.91ms min=103.68ms med=219.42ms max=1.49s    p(90)=338.57ms p(95)=474.16ms
       { expected_response:true }...: avg=268.91ms min=103.68ms med=219.42ms max=1.49s    p(90)=338.57ms p(95)=474.16ms
     http_req_failed................: 0.00%   ✓ 0                   ✗ 26
     http_req_receiving.............: avg=1.84ms   min=0s       med=1.7ms    max=4.44ms   p(90)=4.17ms   p(95)=4.32ms
     http_req_sending...............: avg=104.93µs min=0s       med=0s       max=2.21ms   p(90)=0s       p(95)=384.52µs
     http_req_tls_handshaking.......: avg=0s       min=0s       med=0s       max=0s       p(90)=0s       p(95)=0s
     http_req_waiting...............: avg=266.97ms min=100.12ms med=219.23ms max=1.49s    p(90)=337.72ms p(95)=470.81ms
     http_reqs......................: 26      0.007221/s
     iteration_duration.............: avg=539.06ms min=325.21ms med=388.82ms max=2.01s    p(90)=551.75ms p(95)=1.13s
     iterations.....................: 13      0.003611/s
     vus............................: 0       min=0                 max=1
     vus_max........................: 1       min=1                 max=1


running (1h00m00.4s), 0/1 VUs, 13 complete and 0 interrupted iterations
constantRequestRate ✓ [======================================] 0/1 VUs  1h0m0s  0.00 iters/s
DEBU[3601] Usage report sent successfully
DEBU[3601] Everything has finished, exiting k6 normally!
```

Dannyps commented:
I have added a small utility which can be used to analyze the output of k6.

Example:

```
# node .\result-analyzer\src\main.js .\result.csv 
HTTP Request Metrics Grouped by Route and Type:

 ===

Route: /connect/token
Total Requests: 107
Status Codes:
  200: 107
Timings:
http_req_duration:       524.23 ms
http_req_blocked:        0.02 ms
http_req_connecting:     0.01 ms
http_req_tls_handshaking:        0.00 ms
http_req_sending:        0.06 ms
http_req_waiting:        523.90 ms
http_req_receiving:      0.26 ms
http_req_failed:         0.00 ms

 ===

Route: /api/v1/Challenges
Total Requests: 107
Status Codes:
  201: 107
Timings:
http_req_duration:       290.02 ms
http_req_blocked:        0.01 ms
http_req_connecting:     0.00 ms
http_req_tls_handshaking:        0.00 ms
http_req_sending:        0.02 ms
http_req_waiting:        277.18 ms
http_req_receiving:      12.81 ms
http_req_failed:         0.00 ms
```

tnotheis commented:
> I have added a small utility which can be used to analyze the output of k6.

1. This utility should be part of the script that is used to run the tests, so that you don't have to know/remember to call the script.
2. What does "Type" mean? ("HTTP Request Metrics Grouped by Route and **Type**"

tnotheis commented:
Oh, and 3. what if there are for example `GET /Tokens` and `POST /Tokens`? Currently, your output only shows the rquest path, not the request method.

Dannyps commented:
I've changes the output to group by and include the method:

```
Route: POST /connect/token
Total Requests: 31
Status Codes:
  200: 31
Timings:
http_req_duration:       193.00 ms
http_req_blocked:        0.06 ms
http_req_connecting:     0.04 ms
http_req_sending:        0.03 ms
http_req_waiting:        192.74 ms
http_req_receiving:      0.23 ms
http_req_failed:         0.00 ms

Route: POST /api/v1/Challenges
Total Requests: 31
Status Codes:
  201: 31
Timings:
http_req_duration:       121.26 ms
http_req_blocked:        0.00 ms
http_req_connecting:     0.00 ms
http_req_sending:        0.08 ms
http_req_waiting:        116.09 ms
http_req_receiving:      5.10 ms
http_req_failed:         0.00 ms
```

I have also created windows (PS) and Linux (bash) scripts that take care of compiling and running the test, and then calling the analyzer. Hence,

1. Fixed.
2. I've renamed type to timings
3. Fixed.

tnotheis commented:
I really don't like all the abbreviations you use throughout the code and readme files. This makes it so much harder to understand everything when you work with it for the first time.

Dannyps commented:
@tnotheis I cannot rerequest review here, but it should be good to go.

tnotheis commented on Applications/ConsumerApi/test/ConsumerApi.Tests.Performance/src/libs/backbone-client/fluent-client.ts:
```diff
@@ -0,0 +1,80 @@
+        return this;
+    }
+
+    public asJson(): this {
```
Shouldn't this rather be called seomething like "withJsonBody"? "asJson" looks as if the method would return a json representation of the FluentClient.

tnotheis commented on Applications/ConsumerApi/test/ConsumerApi.Tests.Performance/src/libs/backbone-client/fluent-client.ts:
```diff
@@ -0,0 +1,80 @@
+// @ts-expect-error: k6 uses links to packages, which typescript cannot lint.
+import { Httpx } from "https://jslib.k6.io/httpx/0.1.0/index.js";
+
+export class FluentClient {
```
This class isn't stateless and therefore isn't thread safe. This will become a problem sooner or later.

Dannyps commented on Applications/ConsumerApi/test/ConsumerApi.Tests.Performance/src/libs/file-utils.ts:
I took some metrics:

| Snapshot | Load all entities                                 |
|----------|---------------------------------------------------|
| Light    | 700ms                                             |
| Heavy    | //k6 collapsed after allocating over ~16GB  of RAM |

| Snapshot | Load only identities |
|----------|---------------------------------------------------|
| Light    | 58ms |
| Heavy    | 67901ms |

| Snapshot | Load identities & RelationshipTemplates|
|----------|---------------------------------------------------|
| Light    | 407ms |
| Heavy    | //k6 collapsed after allocating over ~16GB  of RAM |

| Snapshot | Load identities & Relationships|
|----------|---------------------------------------------------|
| Light    | 77ms |
| Heavy    | 67772ms |

I must confess this is not what I was expecting.

Dannyps replied:
loading the data only

tnotheis replied:
Okay, I guess you have things to do then. :)

Dannyps replied:
I'm investigating the possibility of storing the data as json, wondering if the bottleneck here is the parsing of the csv files. Do you have anything else in mind?

tnotheis replied:
You're loading the full file content into memory. I don't know if it is cleaned up once you don't need it anymore (=> once you parsed it and created the corresponding object representations).

Dannyps replied:
Well, me neither. We don't have control over memory management in Javascript. We must also consider that we're using a special engine developed in Go.
It's even possible that the same code would be much more performant when using nodejs, for instance.

tnotheis replied:
Okay, then let's try to avoid the parsing by just saving the JSON. Maybe this already solves our problems.

Dannyps replied:
Before moving to the json representation (it's not that trivial to get), I started taking some timings on the loading process and found a bug. I was actually loading every single relationship template for each identity (like a cross join). I'll be doing further tests to see how much more we can cut on loading times, but I've managed to load all of the the heavy pool in  a blazing fast 109.538s :)

Dannyps replied:
Further optimisations have reduced this time to a bit over 5s on my machine. I'm good with this now.

Dannyps replied:
5500ms for heavy, 230ms for light.

Dannyps commented:
I must call attention to these results: https://github.com/nmshd/backbone/pull/785/files#r1756606368

tnotheis reviewed (changes requested):
I think we're getting there. :) These might actually be the last requested changes.

Dannyps commented on Applications/ConsumerApi/test/ConsumerApi.Tests.Performance/src/libs/data-loader/models.ts:
```diff
@@ -0,0 +1,72 @@
+    ofTypes(...types: string[]): IDataRepresentation;
+}
+
+export class DataRepresentation implements IDataRepresentation {
```
This was called `DREPT`. It was an acronym, so it was renamed to `DataRepresentationForEnmeshedPerformanceTests`. Then we dropped the performance tests part and the Enmeshed. If I recall correctly, you said that it was obvious that this was about performance tests and Enmeshed, hence the change. Now it's so generic 🤷‍♂️

I honestly don't have a better suggestion here. I liked the acronym as a concept itself, once you aquired it, it felt graspable.

tnotheis replied:
How about "Identities"? :)

tnotheis replied:
> If I recall correctly, you said that it was obvious that this was about performance tests and Enmeshed
You are right. And I still think that `ForEnmeshedPerformanceTests` is redundant. My "so generic` was referring to the word "DataRepresentation". What is "Data"? And why is that "Data" not representative enough on its own?

Dannyps replied:
I've changed things around a bit. How about that?

tnotheis replied:
Looks good

Dannyps commented on Applications/ConsumerApi/test/ConsumerApi.Tests.Performance/src/libs/data-loader/models.ts:
```diff
@@ -0,0 +1,72 @@
+     * @param types the names of the pools to be loaded, or simply theIR initial letter(s), e.g.: `.ofTypes("a", "c")` -- loads pools of type a1, a2, a.., c1, c2, c..
+     * @returns
+     */
+    public ofTypes = (...types: string[]): IDataRepresentation => new DataRepresentation(this._pools.filter((pool) => types.some((type) => pool.name.startsWith(type))));
```
We currently support the following pool types/aliases:
 - `e`: never
 - `a1`: app light
 - `a2`: app medium
 - `a3`: app heavy
 - `c1`: connector light
 - `c2`: connector medium
 - `c3`: connector heavy

The idea of this method is to load all pools starting with a certain letter. This comes from the assumption that we'll mostly be loading all a/c pools. If we change to `.withNames()`, and go ahead with the Enum, then a full list must be provided. I don't have an issue with that, but I want to make sure that is the expected change here.

Dannyps commented on Applications/ConsumerApi/test/ConsumerApi.Tests.Performance/src/scenarios/s01.test.ts:
```diff
@@ -0,0 +1,44 @@
+let identityIterator = 0;
+
+export default function (): void {
+    const currentIdentity = testIdentities[identityIterator++];
```
There's a possibility there could be an issue, yes. k6 does not facilitate a solution though. There is https://github.com/olegbespalov/xk6-atomic, but I'd rather avoid custom k6 implementations if possible.

As for the race condition, the truth is that k6 spreads the execution of VUs and respective iterations. Moreover, this being the first line of the default function, I don't expect different threads to become _entangled_. Regardless, I propose a change to further reduce the possibility of complications.

tnotheis commented on Applications/ConsumerApi/test/ConsumerApi.Tests.Performance/src/libs/backbone-client/base-client.ts:
```diff
@@ -0,0 +1,21 @@
+        this.clientId = (__ENV.clientId as string | undefined) ?? "test";
+        this.clientSecret = (__ENV.clientSecret as string | undefined) ?? "test";
+        this.httpxClient = new Httpx({
+            baseURL: defaultBaseUrl,
```
I don't like relying solely on environment variables. I often find it more appealing to use command line arguments. And this wouldn't work if the client reads the configuration from an env var.
Another reason for my suggestion is that if reading the env vars is split across multiple files, it hard to figure out what you can configure and what the names of the env vars are. If you have everything related to configuration in a single file though, this is much easier to figure out.

Dannyps replied:
It is not an env var. It's a cli parameter. It is done under the hood by k6.

tnotheis replied:
Okay, but most of my arguments still apply:
- In order to figure out what arguments exist, you have to scan multiple files.
- You can only configure it using CLI parameters. Using env vars or config files is not possible.

Dannyps replied:
I believe commit f33917d7e19e5535d1a689702b749854c18adebe addresses the issues you raised. 

However, please take into consideration that the `baseUrl` setting is silently overridden when the `--address=<host>:<port>` cli parameter is provided.
